---
title: "Initial analysis"
author: "Yanchao"
output: github_document
---
```{r,warning=FALSE}
library(RColorBrewer)
library(cluster)
library(pvclust)
library(xtable)
library(limma)
library(plyr)
library(lattice)
library(RCurl)
options(download.file.method = "curl")
library(GEOquery)
library(knitr)
library(pheatmap)
  library(stringr)
library(ggplot2)
library(dplyr)
library(reshape2)
library(tidyverse)
```

## Load the data

### geo_GSE18123 data
```{r,warning=FALSE}
 geo_GSE18123 <- getGEO("GSE18123", GSEMatrix = TRUE)
geo_GSE18123<- geo_GSE18123[[1]]
```

### Get expression data of geo_GSE18123
```{r}
#Get expression data  
data_GSE18123<-exprs(geo_GSE18123)
hist(data_GSE18123, col = "gray", main = "GSE70213 - Histogram")

```

It appears a lot of genes have values << 500000. 
We consider taking Log2 transformation.

```{r}
hist(log2(data_GSE18123 + 1), col = "gray", main = "GSE70213 log transformed - Histogram")
log_data_GSE18123<-log2(data_GSE18123 + 1)
log_data_GSE18123<-as.data.frame(log_data_GSE18123)
```

### get melta data of GSE18123
```{r}
## get melta data of GSE18123
prDes_GSE18123 <- pData(geo_GSE18123)[,c("organism_ch1","title",colnames(pData(geo_GSE18123))[grep("characteristics", colnames(pData(geo_GSE18123)))])]
meta_data_GSE18123<-prDes_GSE18123[,1:5]
colnames(meta_data_GSE18123) = c("organism","sample_name","diagnosis","gender","age")
meta_data_GSE18123$diagnosis = as.factor(gsub("diagnosis: ","", meta_data_GSE18123$diagnosis))

meta_data_GSE18123$age = gsub("age: ","", meta_data_GSE18123$age)

meta_data_GSE18123$age<-as.integer(str_extract(meta_data_GSE18123$age, "[0-9]{2,3}"))
meta_data_GSE18123$diagnosis
meta_data_GSE18123$age <- meta_data_GSE18123$age/12

meta_data_GSE18123$diagnosis<-ifelse(meta_data_GSE18123$diagnosis == "PDD-NOS", "AUTISM", ifelse(meta_data_GSE18123$diagnosis == "ASPERGER'S DISORDER", "AUTISM",  ifelse(meta_data_GSE18123$diagnosis == "CONTROL", "CONTROL", ifelse(meta_data_GSE18123$diagnosis == "AUTISM", "AUTISM", "error"))))
meta_data_GSE18123$batch<-"none"

kable(head(meta_data_GSE18123))

```

```{r}
dim(meta_data_GSE18123)
```


```{r}
## convert age to categorial variable
F_meta_data_GSE18123<-meta_data_GSE18123 %>% dplyr::select(organism,sample_name,diagnosis,age,batch)
F_meta_data_GSE18123$age<-ifelse(F_meta_data_GSE18123$age>= 5, "larger or equal to 5", ifelse(F_meta_data_GSE18123$age < 5,"Smaller than 5",  "error"))
F_meta_data_GSE18123$age[is.na(F_meta_data_GSE18123$age)] <- "None"
```

### geo_GSE25507 data
```{r,warning=FALSE}
# 
geo_GSE25507 <- getGEO("GSE25507", GSEMatrix = TRUE)
geo_GSE25507<- geo_GSE25507[[1]]
```

### Get expression data of GSE25507
```{r}
#Get expression data of GSE25507 
data_GSE25507<-exprs(geo_GSE25507)
hist(data_GSE25507, col = "gray", main = "GSE25507 - Histogram")
```

It appears a lot of genes have values < 1000. 
```{r}
hist(log2(data_GSE25507 + 1), col = "gray", main = "GSE25507 log transformed - Histogram")
log_data_GSE25507<-log2(data_GSE25507 + 1)
log_data_GSE25507<-as.data.frame(log_data_GSE25507)
```

### get meta data of GSE25507
```{r}
# get meta data of GSE25507
prDes_GSE25507 <- pData(geo_GSE25507)[,c("organism_ch1","title",colnames(pData(geo_GSE18123))[grep("characteristics", colnames(pData(geo_GSE25507)))])]
meta_data_GSE25507<-prDes_GSE25507[,1:5]
colnames(meta_data_GSE25507) = c("organism","sample_name","batch","diagnosis","age")
meta_data_GSE25507$diagnosis = as.factor(gsub("diagnosis: ","", meta_data_GSE25507$diagnosis))

meta_data_GSE25507$agee = gsub("age: ","", meta_data_GSE25507$age)

meta_data_GSE25507$age<-as.integer(str_extract(meta_data_GSE25507$age, "[0-9]{1}"))
meta_data_GSE25507$diagnosis<-ifelse(meta_data_GSE25507$diagnosis == "group: control", "CONTROL", ifelse(meta_data_GSE25507$diagnosis == "group: autism", "AUTISM", "error"))
meta_data_GSE25507$batch<-ifelse(meta_data_GSE25507$batch == "scan batch: Batch 1", "batch 1", ifelse(meta_data_GSE25507$batch == "scan batch: Batch 2", "batch 2", "error"))
kable(head(meta_data_GSE25507))
```

```{r}
dim(meta_data_GSE25507)

```

```{r}
# convert age to categorial variable
F_meta_data_GSE25507<-meta_data_GSE25507 %>% dplyr::select(organism,sample_name,diagnosis,age, batch)
F_meta_data_GSE25507$age<-ifelse(F_meta_data_GSE25507$age >= 5, "larger or equal to 5", ifelse(F_meta_data_GSE25507$age < 5,"Smaller than 5",  "error"))
F_meta_data_GSE25507$age[is.na(F_meta_data_GSE25507$age)] <- "None"
```

### Combine two meta data
```{r}
## Combine two meta data
Meta_data = rbind(F_meta_data_GSE18123, F_meta_data_GSE25507)
```

Now, we can do some exploratory analysis of the data, for examples, looking at distribution of log gene expression value in gene. Here is a density plot of average log gene expression value in gene in the two datasets.


## density plot
```{r}
# density plot

dat.geneMeans <- c(rowMeans(log_data_GSE25507), rowMeans(log_data_GSE18123)) 
plotDat <- data.frame(mean_gene = dat.geneMeans,
                      Dataset = rep(c('log_data_GSE25507', 'log_data_GSE18123'), each = nrow(log_data_GSE25507)))

(probeAvg <- ggplot(data = plotDat, aes(x = mean_gene, col = Dataset)) +
   geom_density() + 
   ggtitle("Average gene expression value density of two experiments") + 
   xlab("mean of gene ") + 
   ylab("Density") + 
   theme_bw()
)


```

## Quantile normalization
```{r}
# combine data from two experiments into one matrix, each column represents gene expression values of one sample
combine_matrix <- as.matrix(cbind(log_data_GSE18123,log_data_GSE25507))
str(combine_matrix, max.level = 0)
```
```{r}
# quantile normalization
system.time(combine_norm <- normalizeBetweenArrays(combine_matrix))
```
```{r}
dat.geneMeans <- c(rowMeans(combine_norm[, 1:ncol(log_data_GSE18123)]), rowMeans(combine_norm[, ncol(log_data_GSE18123):ncol(combine_norm)])) 
plotDat2 <- data.frame(mean_gene = dat.geneMeans,
                      Dataset = rep(c('log_data_GSE25507', 'log_data_GSE18123'), each = nrow(log_data_GSE25507)))

(probeAvg <- ggplot(data = plotDat2, aes(x = mean_gene, col = Dataset)) +
   geom_density() + 
   ggtitle("Average gene expression value density of two experiments") + 
   xlab("mean of gene ") + 
   ylab("Density") + 
   theme_bw()
)

```

## Plotting gene expression

```{r}
# melt into a format ggplot can easily work with
combine_norm2<-as.data.frame(combine_norm) %>% rownames_to_column("gene")
meltedExpressionMatrix <- combine_norm2 %>% melt(id = "gene") 

head_variable<-head(unique(meltedExpressionMatrix$variable),8)
meltedExpressionMatrix %>% filter(variable==head_variable) %>% 
  ggplot(aes(x = variable, y = value)) +
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))+theme_bw()
```

By looking at the first 8 samples, it shows that the distribution of the sample are similar. 
```{r}
meltedExpressionMatrix %>% filter(variable==head_variable) %>% 
  ggplot(aes(x = value, color = variable)) +
  geom_density() +
  theme_bw()
```

The same result could be shown using the density plot.


## PCA
```{r}
#Scaled data
#scaled_combine_norm = scale(t(combine_norm))
spr_Dat <- t(scale(t(combine_norm)))
```

```{r}
#PCA on the Scaled data
PrinComps = prcomp(spr_Dat, center = FALSE, scale = FALSE)

plot(PrinComps)
```
```{r}
prop<-PrinComps$sdev^2/sum(PrinComps$sdev^2)
barplot(prop[1:10], xlab = 'Principal Components', ylab = 'Proportion of Variance Explained', main = 'Principal Component Analysis')
```

```{r}
# Get the subset of PCs that capture the most variance in your predictors
summary(PrinComps)
```

```{r}
# append the rotations for the first 10 PCs to the metadata
Meta_data_PrinCs = cbind(Meta_data, PrinComps$x[Meta_data$sample_name,1:10])

Meta_data_PrinCs_2 = melt(Meta_data_PrinCs, id=c("organism", "sample_name", "diagnosis", "age","batch"))

```

```{r}
#Relationships between age and PCs
ggplot(data=Meta_data_PrinCs_2, aes(x=age, col=age))+
  facet_wrap(~variable)+
  geom_point(aes(y=value))+
  labs(x="age", y="PC Loading", col="age")+
  theme_bw()
```
```{r}
#Relationships between autism and PCs
ggplot(data=Meta_data_PrinCs_2, aes(x=diagnosis, col=diagnosis))+
  facet_wrap(~variable)+
  theme(axis.text.x = element_text(angle=20))+
  geom_point(aes(y=value))+
  labs(x="diagnosis", y="PC Loading", col="diagnosis")+
  theme_bw()
```

```{r}
#Relationships between batch and PCs
ggplot(data=Meta_data_PrinCs_2, aes(x=batch, col=batch))+
  facet_wrap(~variable)+
  theme(axis.text.x = element_text(angle=20))+
  geom_point(aes(y=value))+
  labs(x="batch", y="PC Loading", col="batch")+
  theme_bw()
```

```{r}
# top 2 PCs related to age
ggplot(data=Meta_data_PrinCs, aes(x=PC1, y=PC2))+
  geom_point(aes(col=age))+
  labs(x="PC1", y="PC2", col="age")+
  theme_bw()
```
```{r}
# top 2 PCs related to autism
ggplot(data=Meta_data_PrinCs, aes(x=PC1, y=PC2))+
  geom_point(aes(col=diagnosis))+
  labs(x="PC1", y="PC2", col="diagnosis")+
  theme_bw()
```
```{r}
# top 2 PCs related to batch
ggplot(data=Meta_data_PrinCs, aes(x=PC1, y=PC2))+
  geom_point(aes(col=batch))+
  labs(x="PC1", y="PC2", col="batch")+
  theme_bw()
```

## Hierarchical Clustering for sample
```{r}
data_to_plot = spr_Dat

# compute pairwise distances
pr.dis <- dist(t(data_to_plot), method = "euclidean")

```

```{r}
pr.hc.a <- hclust(pr.dis, method = "average")

# plot them
plot(pr.hc.a, labels = FALSE, main = "Average", xlab = "")
```
```{r}
# set pheatmap clustering parameters
clust_dist_col = "euclidean"  #‘'correlation'’ for Pearson correlation, ‘'euclidean'’, ‘'maximum'’, ‘'manhattan'’, ‘'canberra'’, ‘'binary'’ or ‘'minkowski'’
clust_method = "average"  #‘'ward.D'’, ‘'ward.D2'’,‘'single'’, ‘'complete'’, ‘'average'’ (= UPGMA), ‘'mcquitty'’ (= WPGMA), ‘'median'’ (= WPGMC) or ‘'centroid'’ (= UPGMC)
clust_scale = "none"  #'column', 'none', 'row'

## the annotation option uses the covariate object (prDes) we defined. It should
## have the same rownames, as the colnames in our data object (data_to_plot).

pheatmap(data_to_plot, cluster_rows = FALSE, scale = clust_scale, clustering_method = clust_method, 
    clustering_distance_cols = clust_dist_col, show_colnames =FALSE, show_rownames = FALSE, 
    main = "Clustering heatmap for combine data", annotation = Meta_data[,c("diagnosis", "age","batch")])
```

## Limma
```{r}
head(spr_Dat %>% as_tibble())
Meta_data_1 = Meta_data
Design = Meta_data_1[,c("diagnosis", "age","batch")]
#Design Matrix
Des_Mat = model.matrix(~ diagnosis+age+batch, Design)
ds_Fit <- lmFit(spr_Dat, Des_Mat)
```

```{r}
ds_EbFit <- eBayes(ds_Fit)
ds_Hits <- topTable(ds_EbFit, coef = grep("diagnosis", colnames(coef(ds_EbFit))), p.value = 0.1, 
    n = Inf)
num_BHhits <- nrow(ds_Hits)

top_Genes <- rownames(ds_Hits)

# Scaled data of topGenes
top_Dat <- spr_Dat[top_Genes, ]

```


```{r}

top_10_Genes <- head(top_Genes,10)
transformGeneExpressionMatrix <- function(expressionMatrix) {
  expressionMatrix <- expressionMatrix %>%
    as.data.frame() %>% 
    column_to_rownames("gene") %>%
    t() %>% as.data.frame() %>% 
    rownames_to_column("sample_id") %>% 
    melt(id = "sample_id") %>% 
    as_tibble() %>% 
    select(sample_id,
           gene = variable, 
           expression = value)
  return(expressionMatrix)
}
Meta_data3 <-Meta_data %>% rownames_to_column("sample_id")
topGenes_ExpressionData <- combine_norm2 %>% 
  filter(gene %in% top_10_Genes) %>%
  transformGeneExpressionMatrix() %>% 
  left_join(Meta_data3, id = "sample_id")

#Examine differences in beta values among top 10 CGIs
ggplot(data=topGenes_ExpressionData, aes(x=diagnosis, y=expression, col=age))+
  geom_point()+
  geom_jitter() +
  stat_summary(aes(y=expression,group=1), fun.y = mean, geom = "line", color = "black")+
  facet_wrap(~gene)+
  labs(x=" diagnosis", y="top 10 gene expression")+
  theme_bw()
```


We start by using clustering algorithms to cluster the top 40 genes that showed differential expression across the different diagnosis stage (BH adjusted p value < 0.1).
```{r}
pheatmap(top_Dat, cluster_rows = TRUE, scale = "none", clustering_method = "average", 
    clustering_distance_cols = "euclidean", clustering_distance_rows = "euclidean", 
    annotation = Meta_data_1[,c("diagnosis", "age","batch")], show_rownames = FALSE,show_colnames  = FALSE)
```

```{r}
geneC_dis <- dist(top_Dat, method = "euclidean")

geneC_hc_a <- hclust(geneC_dis, method = "average")

plot(geneC_hc_a, labels = FALSE, main = "Hierarchical with Average Linkage", xlab = "")
```

```{r}
require(GEOquery)
require(Biobase)
gset <- getGEO("GSE25507", GSEMatrix =TRUE, getGPL=FALSE)
if (length(gset) > 1) idx <- grep("GPL570", attr(gset, "names")) else idx <- 1
gset <- gset[[idx]]

require("biomaRt")
mart <- useMart("ENSEMBL_MART_ENSEMBL")
mart <- useDataset("hsapiens_gene_ensembl", mart)
genesStat <- top_Genes
genesLogReg <- c("1553243_at", "1554675_a_at", "1555097_a_at", "1555591_at", "1556209_at", "1556617_a_at", "1556706_at", "1556786_at", "1557161_at", "1557890_at", "1558857_at", "1561003_at", "1561365_at", "1561578_s_at", "1564010_at", "1569975_at", "202198_s_at", "202933_s_at", "205741_s_at", "206835_at", "207681_at", "208526_at", "209710_at", "209997_x_at", "212836_at", "215636_at", "219196_at", "219758_at", "220183_s_at", "220783_at", "221827_at", "222324_at", "223187_s_at", "226137_at", "228559_at", "230514_s_at", "231446_at", "231487_at", "233379_at", "234110_at", "235104_at", "237597_at", "238314_x_at", "239951_at", "240125_at", "240750_at", "244654_at", "34408_at", "56919_at", "65635_at")
genesSGD <- c("1555707_at", "1555827_at", "1558920_at", "1559470_at", "1561778_at", "1562933_at", "1564052_at", "1567702_at", "1567860_at", "201056_at", "206702_at", "206835_at", "207681_at", "209120_at", "209710_at", "210387_at", "211298_s_at", "213850_s_at", "214882_s_at", "216566_at", "217621_at", "219077_s_at", "220783_at", "223580_at", "224396_s_at", "231842_at", "232752_at", "238314_x_at", "239186_at", "239871_at", "239975_at", "240125_at", "240488_at", "240946_at", "241001_at", "242662_at", "243283_at", "243720_at", "244505_at")

allGenes <- unique(c(genesStat,genesLogReg,genesSGD))

annotLookup <- getBM(mart=mart, attributes=c("affy_hg_u133_plus_2", "ensembl_gene_id", "gene_biotype", "external_gene_name"), filter="affy_hg_u133_plus_2", values=allGenes, uniqueRows=TRUE)

unique(annotLookup[annotLookup$affy_hg_u133_plus_2 %in% genesStat,]$external_gene_name)
# unique(annotLookup[annotLookup$affy_hg_u133_plus_2 %in% genesLogReg,]$external_gene_name)
# unique(annotLookup[annotLookup$affy_hg_u133_plus_2 %in% genesSGD,]$external_gene_name)

# unique(annotLookup[annotLookup$affy_hg_u133_plus_2 %in% intersect(genesLogReg, genesSGD),]$external_gene_name)
# unique(annotLookup[annotLookup$affy_hg_u133_plus_2 %in% intersect(genesLogReg, genesStat),]$external_gene_name)
# unique(annotLookup[annotLookup$affy_hg_u133_plus_2 %in% intersect(genesStat, genesSGD),]$external_gene_name)


```




